tensorflow/core/kernels/scatter_functor.h

```diff
@@ -140,17 +142,44 @@ struct ScatterFunctorBase {
     // indices and params sizes were validated in DoCompute().
     const Index N = static_cast<Index>(indices.size());
     const Index limit = static_cast<Index>(params.dimension(0));
-    for (Index i = 0; i < N; i++) {
-      // Grab the index and check its validity.  An earlier version of the
-      // code checked it and then grabbed it from memory a second time, which
-      // was a security risk since it could have changed in between.
-      const Index index = ::tensorflow::internal::SubtleMustCopy(indices(i));
-      if (!FastBoundsCheck(index, limit)) return i;
-      // Copy last Ndim-1 dimensions of updates[i] to params[index]
-      scatter_op::internal::Assign<op>::Run(params.template chip<0>(index),
-                                            updates.template chip<0>(i));
-    }
-    return -1;
+    auto worker_threads = c->device()->tensorflow_cpu_worker_threads();
+
+    // Calculate the max number of flags.
+    // Based on the shard strategy in work_sharder.cc to limit the flags' size.
+    static const int64 maxFactorPerThread = 10000;
+    Index flags_size = static_cast<Index>(maxFactorPerThread * worker_threads->num_threads);
+    flags_size = N <= flags_size ? N : flags_size;
+    // use atomic_flag and spin lock for multiple thread.
+    auto flags = std::unique_ptr<std::atomic_flag[]>(
+            new std::atomic_flag[flags_size]{ATOMIC_FLAG_INIT});
+
+    // store the result.
+    mutex mu;
+    Index result = -1 GUARDED_BY(mu);
+    auto work = [&params, &updates, &indices, &flags, flags_size, limit, &mu, &result]
+            (int64 start, int64 end) {
+      Index flags_idx;
+      for (Index i = start; i < end; i++) {
```

Comment:
There is now a lot of code that gets duplicated, can you factor out the loop since only the function is changed in the three occurrences.  And maybe also the code that creates the flags since it is also duplicated twice.